### PR TITLE
feat: create-notifications

### PR DESCRIPTION
**Pull request by CFCIfe**
## What does the PR do?

This PR implements the creation of notifications for users. The new endpoint allows users to create notifications based on specific criteria and preferences. It handles various scenarios including authentication, validation, and error responses.

**Features Implemented:**

- **Notification Creation:** Allows authenticated users to create notifications.
- **Error Handling:** Manages errors for unauthenticated users, invalid data, and other edge cases.

## How should this be manually tested?

1. **Login to a user account and obtain an access token.**
   <br />

2. **Update the notification_settings table with sample data for the created user.**:

> Make sure the user ID matches the ID of the logged-in user.
> Make sure the email_notifications column are set to true.
> Configure a simple SMTP server on mailtrap.io to receive email notifications.

```sql
INSERT INTO "public"."notification_settings" ("id", "created_at", "updated_at", "user_id", "email_notification_activity_in_workspace", "email_notification_always_send_email_notifications") VALUES
('e5fbc63e-d026-4e1a-99c0-864d1b65a793', '2024-07-28 18:18:54.823437', '2024-07-28 18:21:54.423651', 'e5fbc63e-d026-4e1a-99c0-864d1b65a793', 'TRUE', 'TRUE');
```

<br />

3. **Create a Notification:**

   - Make a POST request to the `/api/v1/notifications` endpoint with the following payload:

   ```json
   Authorization: Bearer <access_token>
   {
     "message": "Welcome to HNGi8 user"
   }
   ```

   - **Successful Response:** Upon successful creation, the API returns a 201 CREATED status code.

   ```json
   {
     "status": "success",
     "status_code": 201,
     "message": "Notification created successfully",
     "data": {
       "notifications": [
         {
           "notification_id": "c90ec520-bdf8-4629-a9f4-098aa782b409",
           "user_id": "e5fbc63e-d026-4e1a-99c0-864d1b65a793",
           "message": "welcome again",
           "is_read": false,
           "created_at": "2024-07-30T08:21:30.558Z"
         }
       ]
     }
   }
   ```

image

- **Authentication Error:**

  - If the user is not authenticated, the following error is returned:

  ```json
  {
    "status": "error",
    "message": "User is currently unauthorized, kindly authenticate to continue",
    "status_code": 401
  }
  ```

  No Token Error

- **Invalid UserID Error:**

  - If the user ID is invalid, the response will be:

  ```json
  {
    "status": false,
    "status_code": 400,
    "error": "Not found",
    "message": "An error occurred"
  }
  ```

  Invalid User ID Error

- **Validation Errors:**

  - If the input data is invalid or incomplete, a 400 Bad Request status is returned with validation error messages.

  ```json
  {
    "status": false,
    "status_code": 400,
    "error": "Bad Request",
    "message": "message must be a string, message should not be empty"
  }
  ```

  Validation Error

## Task Checklist:

- [x] Create a POST endpoint `/api/v1/notifications` to create notifications for logged-in users.
- [x] Implement server-side logic for handling notification creation.
- [x] Ensure proper authentication and authorization checks.
- [x] Update notification details in the database.
- [x] Write tests for the notification creation process.
- [x] Update API documentation to reflect the new endpoint and its functionality.
- [x] Send emails if user opts-in for email notifications

### Unit Tests

- [x] It should create a notification successfully.
- [x] It should throw BadRequestException if user not found.
- [x] It should handle internal server errors.
- [x] It should throw NotFoundException with specific user_id not found.

image

## Reference Issue:

[Create Notifications](https://linear.app/team-bingo/issue/BAC2-41/create-notifications)

## INFO FOR THE DEVOPS

A new secret: `SUPPORT_EMAIL=email` must be added to the `.env`

### CODE REVIEW

File: src/modules/notifications/dto/create-notification-response.dto.ts [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+  message: string;
+  is_read: boolean;
+  created_at: string | Date;
+}
+
+class Data {
+  notifications: ResponseNotification[];
+}
+
+export class CreateNotificationResponseDto {
```
Comment (Cobidex) [inline src/modules/notifications/dto/create-notification-response.dto.ts:15]: Why validators on a response Dto?

**Review by alexindevs — DISMISSED** [review]
LGTM!

File: src/modules/notifications/dto/create-notification-response.dto.ts (hunk 1 above)
Comment (CFCIfe) [reply src/modules/notifications/dto/create-notification-response.dto.ts:15]: oversight lol. Removed

**Review by Cobidex — DISMISSED** [review]

**Review by Heba-WebDev — APPROVED** [review]

**Review by davidolanrewaju — APPROVED** [review]

**Review by davidolanrewaju — DISMISSED** [review]

File: src/modules/notifications/notifications.service.ts [hunk 2]
```diff
@@ -105,4 +119,106 @@ export class NotificationsService {
...
+        status: 'error',
+        error: 'Not Found',
+        status_code: HttpStatus.NOT_FOUND,
+      });
+    }
+
+    return user;
+  }
+
+  private async getNotificationSettingsById(user_id: string): Promise<NotificationSettings> {
```
Comment (Cobidex) [inline src/modules/notifications/notifications.service.ts:171]: Rename this to getNotificationSettingsByUserId for more clarity

File: src/modules/notifications/notifications.service.ts [hunk 3]
```diff
@@ -105,4 +119,106 @@ export class NotificationsService {
...
+    try {
+      return await this.notificationSettingsService.findNotificationSettingsByUserId(user_id);
```
Comment (Cobidex) [inline src/modules/notifications/notifications.service.ts:173]: No need for await here, function is returning and will wait.

**Review by Heba-WebDev — DISMISSED** [review]

**Review by Cobidex — DISMISSED** [review]
LGTM

**Review by rugue — DISMISSED** [review]

**Review by Cobidex — DISMISSED** [review]
LGTM

**Review by rugue — DISMISSED** [review]

**Review by rugue — APPROVED** [review]

**Review by rugue — APPROVED** [review]

**Review by Homoakin619 — APPROVED** [review]

**Review by davidolanrewaju — APPROVED** [review]

**Pull request merged**